packages/gatsby/src/query/index.js

```diff
@@ -1,152 +1,33 @@
-          dirtyIds.add(n)
-        }
-      })
-    }
-
-    // Find connections that depend on this node so are now invalid.
-    if (state.componentDataDependencies.connections.has(node.internal.type)) {
-      state.componentDataDependencies.connections
-        .get(node.internal.type)
-        .forEach(n => {
-          if (n) {
-            dirtyIds.add(n)
-          }
-        })
-    }
-
-    return dirtyIds
-  }, new Set())
-
-  boundActionCreators.deleteComponentsDependencies([...uniqDirties])
-
-  queuedDirtyActions = []
-  return uniqDirties
-}
-
-const popNodeAndDepQueries = state => {
-  const nodeQueries = popNodeQueries(state)
-
-  const noDepQueries = findIdsWithoutDataDependencies(state)
-
-  return _.uniq([...nodeQueries, ...noDepQueries])
-}
 
 /**
- * Calculates the set of dirty query IDs (page.paths, or
- * staticQuery.hash's). These are queries that:
- *
- * - depend on nodes or node collections (via
- *   `actions.createPageDependency`) that have changed.
- * - do NOT have node dependencies. Since all queries should return
- *   data, then this implies that node dependencies have not been
- *   tracked, and therefore these queries haven't been run before
- * - have been recently extracted (see `./query-watcher.js`)
+ * Calculates the set of dirty query IDs (page.paths, or staticQuery.id's).
  *
- * Note, this function pops queries off internal queues, so it's up
- * to the caller to reference the results
- */
-
-const calcDirtyQueryIds = state =>
-  _.union(popNodeAndDepQueries(state), popExtractedQueries())
-
-/**
- * Same as `calcDirtyQueryIds`, except that we only include extracted
- * queries that depend on nodes or haven't been run yet. We do this
- * because the page component reducer/machine always enqueues
- * extractedQueryIds but during bootstrap we may not want to run those
- * page queries if their data hasn't changed since the last time we
- * ran Gatsby.
+ * Dirty state is tracked in `queries` reducer, here we simply filter
+ * them from all tracked queries.
  */
-const calcInitialDirtyQueryIds = state => {
-  const nodeAndNoDepQueries = popNodeAndDepQueries(state)
-
-  const extractedQueriesThatNeedRunning = _.intersection(
-    popExtractedQueries(),
-    nodeAndNoDepQueries
-  )
-  return _.union(extractedQueriesThatNeedRunning, nodeAndNoDepQueries)
+const calcDirtyQueryIds = state => {
+  const { trackedQueries, trackedComponents } = state.queries
+
+  const queriesWithBabelErrors = new Set()
+  for (const component of trackedComponents.values()) {
+    if (hasFlag(component.errors, FLAG_ERROR_BABEL)) {
+      for (const queryId of component.pages) {
+        queriesWithBabelErrors.add(queryId)
+      }
+    }
+  }
+  // Note: trackedQueries contains both - page and static query ids
+  const dirtyQueryIds = []
+  for (const [queryId, query] of trackedQueries) {
+    if (query.dirty > 0 && !queriesWithBabelErrors.has(queryId)) {
+      dirtyQueryIds.push(queryId)
+    }
+  }
+  return dirtyQueryIds
```

Comment:
Figure out which queries to actually run based on the tracked dirty state. Also, skipping queries that relate to components with errors.